### PR TITLE
Feature mingw64 changes

### PR DESCRIPTION
**Pull request by guy-rouillier**
Changes needed to compile windoze/tvision under Windows 7 64-bit using mingw-w64 64-bit compiler and MSYS2 64-bit.  Unfortunately, some files, such as wingrdis.cc, have non-ASCII characters embedded, and those got replaced by the CodeLite editor.  Ignore those changes.

Also see the comments on the 3 commits.  There is still work to do to ensure the changes remain compatible across 32-bit and 64-bit platforms.  I think that particular problem is limited to regex.c.  If we can't replace regex.c with std lib equivalents, then perhaps we can convert all those macros to functions to solve the major issues.

### CODE REVIEW

**Comment by guy-rouillier** [comment]
For the 2nd commit, the only file that required me to use 64-bit explicit changes was regex.c.  However, after posting these commits, I figured out that regex.c is not used by either the library or any of the examples.  GNU says here [https://www.gnu.org/software/regex/](https://www.gnu.org/software/regex/) that this code "has long been decommissioned", and they recommend changing to newer standardized regular expressions.  So, I suggest simply removing regex.c from the project.  The remainder of the changes should be compatible across platforms and compilers.

Sorry about the whitespace changes.  They did not show when I submitted the commits.

File: classes/tfilelis.cc [hunk 1]
```diff
@@ -342,7 +342,7 @@ void TFileList::readDirectory( const char *aWildCard )
   else
     {
      slash = wildcard;
-     path = ".";
+     path = (char*) ".";
```
Comment (set-soft) [inline classes/tfilelis.cc:345]: Why this? Are you sure?
This is converting the constant into something writable. A very bad idea.
I think this line should be removed and the strcpy that uses it should be:
```
strcpy(dirpath,".");
```

File: classes/wingr/wingrscr.cc [hunk 2]
```diff
@@ -37,7 +37,6 @@
 #define Uses_string
 #define Uses_stdlib
 #define Uses_unistd
-#define Uses_limits
```
Comment (set-soft) [inline classes/wingr/wingrscr.cc:40]: You are removing a line added by @windoze 
It seems to be needed for some Windows compilers. Most probably MSVC.

File: classes/wingr/wingrscr.cc [hunk 3]
```diff
@@ -165,25 +164,25 @@ int TScreenWinGr::GetFontGeometryRange( unsigned &wmin, unsigned &hmin
 
 /*
  *  Time callback to blink the cursor
- */ 
+ */
 VOID CALLBACK cursorProc( HWND hwnd          // handle of window for timer messages
                         , UINT uMsg          // WM_TIMER message
-                        , UINT_PTR idEvent       // timer identifier
+                        , UINT idEvent       // timer identifier
```
Comment (set-soft) [inline classes/wingr/wingrscr.cc:170]: You are reverting a change by @windoze 
According to Win32 API the @windoze change is correct.

File: classes/wingr/wingrscr.cc [hunk 4]
```diff
@@ -165,25 +164,25 @@ int TScreenWinGr::GetFontGeometryRange( unsigned &wmin, unsigned &hmin
...
+    }
   }
-  
-  SetTimer( hwnd                 /* Reload timer        */ 
+
+  SetTimer( hwnd                 /* Reload timer        */
           , idEvent
           , TScreenWinGr::cursorDelay / 1000   /* from usecs to msecs */
-          , cursorProc );
+          , (TIMERPROC) cursorProc );
```
Comment (set-soft) [inline classes/wingr/wingrscr.cc:185]: This shouldn't be needed if you don't break the code at line 170.

File: classes/wingr/wingrscr.cc [hunk 5]
```diff
@@ -597,9 +596,10 @@ int TScreenWinGr::SetDisPaletteColors( int from
   number= 0;                 /* Now is a counter */
 
   while( from < to )
-  { colorMap[ from++ ]= RGB( colors[ from ].R
+  { colorMap[ from ]= RGB( colors[ from ].R
                            , colors[ from ].G
                            , colors[ from ].B );
+    from++;
```
Comment (set-soft) [inline classes/wingr/wingrscr.cc:602]: Is that for clarity?

File: classes/wingr/wingrscr.cc [hunk 6]
```diff
@@ -734,13 +734,13 @@ int TScreenWinGr::System( const char * command
   }
 
   if ( pidChild )                      // Leave in the background
-  { return( *pidChild= (int)pi.hProcess );  // Give process identifier
+  { return( *pidChild= (_pid_t) pi.hProcess );  // Give process identifier
```
Comment (set-soft) [inline classes/wingr/wingrscr.cc:737]: Why `_pid_t`? The argument is `pid_t`

File: classes/wingr/wingrscr.cc [hunk 7]
```diff
@@ -756,7 +756,7 @@ TScreen * TV_WinGrDriverCheck()
 
   if ( !TScreen::initialized )
   { delete drv;
-    drv= NULL;
```
Comment (set-soft) [inline classes/wingr/wingrscr.cc:759]: OMG!!! don't get me wrong but your editor is a nightmare.
Why is it adding spaces all over the place? and also destroying the encoding?
It was really hard to read this patch.

File: compat/regex.c [hunk 8]
```diff
@@ -4109,6 +4110,9 @@ re_match_2 ( struct re_pattern_buffer *bufp, const char *string1,
             unsigned dummy_low_reg, dummy_high_reg;
             unsigned char *pdummy;
             const char *sdummy;
+            
+            UNUSED(sdummy);
+            UNUSED(pdummy);
```
Comment (set-soft) [inline compat/regex.c:4115]: Assuming any compiler has UNUSED is not safe

File: compat/releasec.c [hunk 9]
```diff
@@ -24,7 +24,7 @@ extern void __tvWin32Yield(int micros);
 
 void CLY_ReleaseCPU()
 {
- __tvWin32Yield(-1);
+ __tvWin32Yield(100); // -1 guyr
```
Comment (set-soft) [inline compat/releasec.c:27]: What about using the Sleep(100) found in the else?

**Comment by set-soft** [comment]
> However, after posting these commits, I figured out that regex.c is not used by either the library or any of the examples.

It doesn't mean you can blantly remove it. You must understand that this port of TV was designed to be used by RHIDE and SETEdit. Some features are there just to make those programs easier to maintain.

**Comment by set-soft** [comment]
I'm incorporating most of your patches to the original Source Forge code.

**Comment by set-soft** [comment]
Hi @windoze !

I think I confused you. This pull-request was created by @guy-rouillier 
It contains valuable patches for MinGW 64 bits target.
**But**
1) He took as reference code older than your HEAD. So it reverts some of your patches. I confirmed it looking at VOID CALLBACK cursorProc https://github.com/windoze/tvision/blob/610ce5bc797a3c122debddbdd726d43a9ad300d1/classes/wingr/wingrscr.cc#L170
As you correctly fixed in windoze/tvision@63090519575bbecaf5d2a5a326cd67109527e5ad this must be `UINT_PTR`
2) He introduced a lot of random whitespaces
3) Some non-UTF-8 characters were broken

Today I moved my repo from Source Forge to GitHub and you can see 8 patches I picked from this PR (they are repeated because I imported the repo twice, GitHub reported errors, but it turned out to be a delay in the import).

I suggest you to revert this PR and apply the patches individually.

I commented various parts of the PR pointing problems.

BTW I also have problems with notifications and I don't see what exactly is needed to get all the notifications. This why I like to mention people in my replies.

**Comment by windoze** [comment]
Be honest, I started this repos just because there was an project needed the TV lib to be built with CMake and C++11 compilers, I never looked too deep into the source code, and never had enough time to maintain it.
Given the situation that you, the original author, and @guy-rouillier, both have excellent progress, I think it's a good time to end my fork.
I'll put this repo into the archive state and you may make your own developments without my obstructions.

Thanks for your support.

**Comment by set-soft** [comment]
Ok, thanks for sharing your patches @windoze !

**Comment by guy-rouillier** [comment]
> > However, after posting these commits, I figured out that regex.c is not used by either the library or any of the examples.
> 
> It doesn't mean you can blantly remove it. You must understand that this port of TV was designed to be used by RHIDE and SETEdit. Some features are there just to make those programs easier to maintain.

I'm surprised to be receiving these updates after 2 years.  Thank you for the effort to incorporate my changes.  I will do a fresh clone and see what the state of compatibility with a current version of mingw-w64 is like.  This will take me some time, as I've moved on to other projects since 2018.  Looks like you are currently maintaining this code on GitHub.

Regarding regex.c, if it is not used by the tvision project and is causing compatibility issues, wouldn't a simple solution be to remove it, and then let any projects that reference tvision add whatever regex capabilities they require?  At this point, regex is a standard feature of most compilers, so regex.c likely is no longer required.

**Pull request merged**